L44-python-rich-status.md

```diff
@@ -0,0 +1,474 @@
+            details=[detail]
+        )
+
+        context.set_code(rich_status.code)
+        context.set_details(rich_status.message)
+        context.set_binary_status(rich_status.serializeToString())
+        ...
+```
+
+#### Pros
+* Similar to C++'s strictness: code/message/details are independent
+* Promised developers the usage of metadata entry `grpc-status-details-bin`
+
+#### Cons
+* Verbosity
+* No code/message validation at all
+* Confusing definition of message/details
+
+
+### Option 6: Pack as Tuple & Splat if Needed
+
+The new API will be added to [ServicerContext](https://grpc.io/grpc/python/grpc.html#service-side-context), and named `set_status`. It accepts two positional arguments and a keyword argument, so setting the status code and status message is mandatory but not details. At the same time, current API `set_code`/`set_details` will be labeled as "deprecated".
+
+The function signature should be:
+
+```Python
+def set_status(code, message, details=""): pass
+```
+
+The usage should look like:
+
+```Python
+# Server side
+from grpc_status import rpc_status
+from google.protobuf import any_pb2
+
+def ...Servicer(...):
+    def ARPCCall(request, context):
+        ...
+        detail = any_pb2.Any()
+        detail.Pack(
+            rpc_status.error_details_pb2.DebugInfo(
+                stack_entries=traceback.format_stack(),
+                detail="Can't recognize this argument",
+            )
+        )
+        rich_status = grpc_status.status_pb2.Status(
+            code=grpc_status.code_pb2.INVALID_ARGUMENT,
+            message='API call quota depleted',
+            details=[detail]
+        )
+
+        context.set_status(*rpc_status.convert(rich_status))
+```
+
+#### Pros
+* Concise API call
+* Avoid introducing new classes
+
+#### Cons
+* Recommend using splat is not Pythonic
+* Need to educate developers about the priority of those three APIs
+
+
+### Why we need a new package `grpcio-status`?
+
+Although we can't enforce the consistency of status code, status message, and status details, there is still value for the new package to help developers validate that. It demonstrates our recommendation through the design.
+
+One more reason is that in the future, the `google.rpc.status.Status` may not be the only status proto we accept. This package provides a straightforward mapping between different status proto message.
+
+
+### Not Swapping the Two Concepts
+
+Ideally speaking, the **status message** and **status details** are misused in Python implementation, and this proposal is our best chance to get them right. According to spec, the correct definition should be as followed:
+
+* **Status message** should be interpreted as a text (Unicode allowed) field that indicates the status of the RPC call, and it is intended to be read by the developers.
+* **Status details** should be understood as a encoded `google.rpc.status.Status` proto message that serves as rich status error details about the RPC Call.
```

Comment:
There is not actually a spec for this, is there? I would just simplify this section by saying what Python calls 'details' actually corresponds to `grpc-message`, and that this discrepancy is unfortunately deeply embedded in the gRPC Python API and unlikely to be changed.

Reply:
The specs for both "grpc-message" and "grpc-status-details-bin" fields are not directly describing their usage, and you are right we shouldn't do the definition for them in this gRFC. The last section is rewrote into:

> ### Semantic difference of `details` between gRPC Python and `Status` proto

> gRPC Python uses the field name `details` to store data that will later be transmitted in `grpc-message` header. But the `details` field in proto `Status` is a list for error details proto messages. Unfortunately, this discrepancy is deeply embedded in the gRPC Python API and unlikely to be changed.